README
ABOUT MPG321
------------

mpg321 is a Free replacement for mpg123, a very popular command-line mp3
player. mpg123 is used for frontends, as an mp3 player and as an mp3 to
wave file decoder (primarily for use with CD-recording software.) In all 
of these capacities, mpg321 can be used as a drop-in replacement for
mpg123.

mpg321 is a project conceived and written by Joe Drew <hoserhead@woot.net>.
It uses code from the MAD distribution by Rob Leslie; XMMS; and plaympeg
from the SMPEG distribution. mpg321 is licensed under the GNU General
Public License version 2, as documented in the 'COPYING' file.

WHY MPG321?
-----------

mpg123 is very popular because of its relative simplicity and its general
facility. It is very good at what it does. It is, however, under a non-free
license: you can't incorporate its code into your own without getting 
special permission, and commercial entities also need special permission
just to use it.

Being a Free Software advocate, I saw a deficiency in the dependency many
people had on mpg123. Therefore, I created mpg321, and as more features
of mpg123 are implemented, it becomes less and less likely that anyone
will need mpg123 for anything any more.

INSTALLING MPG321
-----------------

Please read the document 'INSTALL' to find mpg321's dependencies and
configuration, compilation and installation instructions.

mpg321 has been tested on Debian GNU/Linux (x86), as well as Sun Solaris
(UltraSparc). It is also available from several BSD vendors, and may be 
included in other Linux distributions. Any problems encountered on any 
system, though, should be reported to the author.

GETTING INVOLVED/INFORMED
-------------------------

If you like mpg321, please donate to the Electronic Frontier Foundation (EFF).
They help ensure that software for any purpose can continue to be written 
legally. Find out more information at http://www.eff.org .

To be informed of any new releases of mpg321, please subscribe to the
mpg321-announce mailing list: send an email with 'subscribe' in the body or 
subject to mpg321-announce-request@lists.sourceforge.net, or visit
http://lists.sourceforge.net/lists/listinfo/mpg321-announce .

There are a number of tasks that are still on-going with the develompent
of mpg321. If you're interested, please read the TODO and HACKING files.
Also, always feel free to e-mail me, Joe Drew <hoserhead@woot.net>.

There is also a sourceforge project for mpg321:
http://sourceforge.net/projects/mpg321

HISTORY OF MPG321
-----------------

`mpg321' started long ago (sometime in 1999) as a wrapper script for 
the mpg123 frontend of the media player FreeAmp. It was contributed 
to the Debian package of FreeAmp, and for several revisions was 
included as 'mpg123.freeamp.' (This script has since been removed.)

mpg123.freeamp didn't work very well, and was limited by FreeAmp's 
implementation of the mpg123 frontend, but it worked for what I needed
it to do: play music from the command-line without needing mpg123
installed.

Early in 2001, I decided that a real alternative to mpg123 was needed,
and started looking around for a good mp3 decoding library. I came
across SMPEG, which was written by Loki Software, and based on SDL.
It was very simple programatically and handled most everything that
was needed. mpg321 was born, and it did all the simple things that
were required of it.

SMPEG, however, required SDL for output, which in turn required many
other libraries. This was a very sub-optimal arrangement. So, in 2001,
I performed what amounted to almost a full rewrite of mpg321, porting
it to Rob Leslie's MAD MPEG audio decoder library. I chose MAD because
it provides very high-quality output, and also because it operates entirely 
with fixed point (integer) instructions. 

At the same time, I also switched to the Ogg Vorbis (Xiphophorous) project
libao for all audio output.

That's pretty much where mpg321 stands right now. There are still
unimplemented features, and undoubtedly still bugs. But it works well
for many people, and it liberates one more dependency people had on
non-Free software. That's worth it, to me.


debian/changelog
mpg321 (0.3.2-1) unstable; urgency=low

  * Fixed bugs to stdout-output (remove garbage in buffer from wrong 
    stdout fprintf's, it should have been stderr instead)
  * Fixed bug when closing the stream of AO device that caused wrong
    header to the output buffer for the normal operation. Buffered output
    has no problem with that.

 -- Nanakos Chrysostomos <nanakos@wired-net.gr>  Sun, 25 Mar 2012 14:27:09 +0300

mpg321 (0.3.1-1) unstable; urgency=low
 
  * Fixed debian/rules, configure.ac, volume.c, mpg321.c and remote.c
    so that mpg321 can compile to architectures that do not support 
    ALSA.
  * Fixed typo ("lighweight") in the man page.
    Thanks to A.Costa <agcosta@gis.net> (Closes: Bug#650150)

 -- Nanakos Chrysostomos <nanakos@wired-net.gr>  Sun, 19 Feb 2012 13:00:36 +0300

mpg321 (0.3.0-1) unstable; urgency=low

  * Fixed trailing "/" when printint directory. 
    Bug reported from Erik (Gentoo).
  * Fixed mistake for '--cdr' option. It should be 'cdr file'
    than 'wave file' in output.
  * mpg321 now supports multiprocessing buffering.Check '-b' option. 
    (Closes: Bug#113405).
  * Added '-3' or '--restart' option in man file.
  * Added ALSA volume control when using output buffer.
  * Added Mute/unmute into Basic Keys functionality.
  * Fixed -Werror=format-security issues.
    Thanks to Micah Gersten <micahg@ubuntu.com>. (Closes: Bug#646317)

 -- Nanakos Chrysostomos <nanakos@wired-net.gr>  Sun, 13 Nov 2011 13:42:36 +0300

mpg321 (0.2.13-4) unstable; urgency=low

  * Fixed bug when trying to "load file" from a remote instance and 
    an error was raised repeatdly until mpg321 process died with 
    "too many open files". (Closes: Bug#128676)
  * Added '-3' or '--restart' option that handles non existent or 
    corrupted files more elegantly when in remote mode. When option 
    is enabled the remote shell does not die and is restarted waiting
    for a new command.
  * Fixed FTBFS for mpg321 by removing debian_changes patch file in 
    debian/patches directory. (Closes: Bug#643222).

 -- Nanakos Chrysostomos <nanakos@wired-net.gr>  Thu, 29 Sep 2011 19:33:24 +0300

mpg321 (0.2.13-3) unstable; urgency=low

  * Fixed the really annoying bug which didn't restore the TTY 
    when mpg321 exited.
  * Updated S-V to 3.9.2 
  	- debian/copyright file has been revised.
  	- debian/rules file has been revised.
  * Fixed '-a' option to conform with ALSA's uniform device naming. Also
    changed man to conform with this change. Now you can use alsa driver
    by setting the name of the audio device using the hw:x,y syntax, 
    where x and y are numbers. For example, if there is only one device 
    installed, in most cases, the device should be named hw:0,0. 
    When there is only one device, the device should always have the 
    same name and numbers. Finally if the default driver which mpg321
    is compiled is ALSA then you don't have to specify it with the '-o'
    option and you can use '-a' only to specify the audio device.
    (Closes: Bug#623941).
  
 -- Nanakos Chrysostomos <nanakos@wired-net.gr>  Sun, 24 Jul 2011 15:08:09 +0300

mpg321 (0.2.13-2) unstable; urgency=low

  * Added '-K' option so the user can enable/disable the Basic Keys
    functionality. This also resolves the 100% CPU problem when mpg321 
    is run by asterisk for music hold on.
    Thanks to James Bottomley for pointing that out. (Closes: Bug#619773).     
  * Added new algorithm when playing files randomly until interrupted. The
    advertised behavior is much better than the previous one.Patch from 
    Peter Selinger.Thanks.

 -- Nanakos Chrysostomos <nanakos@wired-net.gr>  Tue, 05 Apr 2011 23:23:54 +0200

mpg321 (0.2.13-1) unstable; urgency=low

  * Added uClinux support for embedded systems (Blackfin, Atmel and others).
  * Added some Basic Control Keys, Volume Up/Down & Skip, for the moment.
  * Updated S-V to 3.9.1
       - debian/copyright file has been revised.
  * If mmap() fails we fall back to read(). (Closes: Bug#148971).
    (Closes: Bug#196787)
  * Removed built-in getopt function.
  * Problem solved when -x option was invoked and mpg321 couldn't obtain the
    current terminal file. If none is forthcoming (e.g. xterm window ops are
    disabled) then mpg321 does not block.
  * Removed the dependency for the libaudio-scrobbler-perl package from
    debian/control file. (Closes: Bug#588051).
  * Added LDFLAGS += -Wl,-z,defs -Wl,--as-needed in debian/rules file
    to limit the dependencies a bit. Dropped '-c' option from 'install'
    command. Thanks to Sven Hoexter <sven@timegate.de>.

 -- Nanakos Chrysostomos <nanakos@wired-net.gr>  Sun, 20 Feb 2011 15:35:00 +0300

mpg321 (0.2.12-1) unstable; urgency=high

  * New upstream release.
  * Added AudioScrobbler support.
  * Added FFT analysis support on PCM data for Remote mode play.
  * mpg123 has been relicensed under the LGPL/GPL. Changed man page
    accordingly. (Closes: Bug#533674).
  * mpg321 now exits code 1 after failing to play music. (Closes: Bug#501334).
  * Fixed ID3v2 tag problem which resulted in a splat-sound at the start of
    the file. Patch from Martin Wellard. Thanks.
  * Added support for recursive reading of directories. Patch from Giuseppe
    Scrivano <gscrivano@gnu.org> and Michal Čihař <nijel@debian.org>.
    (Closes: Bug#266887).
  * Updated S-V to 3.9.0
  	- debian/copyright file is now UTF-8 encoded.

 -- Nanakos Chrysostomos <nanakos@wired-net.gr>  Tue, 29 Jun 2010 21:44:36 +0300

mpg321 (0.2.11-3) unstable; urgency=high

  * Now mpg321 displays the right MPEG version. (Closes: Bug#197726).
  * New function added for restoring the correct xterm title.
  * New troubleshooting information added for the resources that mpg321
    uses for xterm control.
  * Bug #182122 should close now because in this version of Debian we don't
    face such problems. So I am closing this bug. (Closes: Bug#182122).
  * Man page now includes the new "-l" option. (Closes: Bug#575836).
    Thanks to Memnon Anon.

mpg321.h
/*
    mpg321 - a fully free clone of mpg123.
    Copyright (C) 2001 Joe Drew
    Copyright (C) 2006, 2007, 2008, 2009, 2010, 2011, 2012 Nanakos Chrysostomos
    
    Originally based heavily upon:
    plaympeg - Sample MPEG player using the SMPEG library
    Copyright (C) 1999 Loki Entertainment Software
    
    Also uses some code from
    mad - MPEG audio decoder
    Copyright (C) 2000-2001 Robert Leslie
    
    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#ifndef _MPG321_H_
#define _MPG321_H_

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <sys/types.h>
#include <stdio.h>
#include <limits.h>
#include <ao/ao.h>
#include <mad.h>
#include <semaphore.h>


#define FAKEVERSION "0.3.0-1"
#define VERSIONDATE "2011/10/08"

#ifndef PATH_MAX
#define PATH_MAX	4096
#endif

/* playlist structure */
typedef struct pl
{
    char **files;
    int numfiles;
    int files_size;
    int random_play;
    char remote_file[PATH_MAX];
} playlist;

/* Private buffer for passing around with libmad */
typedef struct
{
    /* The buffer of raw mpeg data for libmad to decode */
    void * buf;

    /* Cached data: pointers to the dividing points of frames
       in buf, and the playing time at each of those frames */
    void **frames;
    mad_timer_t *times;

    /* fd is the file descriptor if over the network, or -1 if
       using mmap()ed files */
    int fd;

    /* length of the current stream, corrected for id3 tags */
    ssize_t length;

    /* have we finished fetching this file? (only in non-mmap()'ed case */
    int done;

    /* total number of frames */
    unsigned long num_frames;

    /* number of frames to play */
    unsigned long max_frames;

    /* total duration of the file */
    mad_timer_t duration;

    /* filename as mpg321 has opened it */
    char filename[PATH_MAX];
    
    /* the playlist. */
    playlist *pl;
} buffer;

typedef struct
{
    int opt;
    char *devicetype;
    char *device;
    signed long seek;
    signed long maxframes;
    mad_fixed_t volume;
    int skip_printing_frames;
} mpg321_options;    

extern mpg321_options options;
extern ao_device *playdevice;
extern mad_timer_t current_time;
extern unsigned long current_frame;
extern int stop_playing_file;
extern int shuffle_play;
extern char *playlist_file;
extern int quit_now;
extern char remote_input_buf[PATH_MAX + 5];
extern int file_change;
int loop_remaining;

extern int status;
extern int scrobbler_time;
extern char *scrobbler_args[6];

enum
{
    MPG321_STOPPED       = 0x0001,
    MPG321_PAUSED        = 0x0002,
    MPG321_PLAYING       = 0x0004,
    MPG321_REWINDING     = 0x0008,
    MPG321_SEEKING       = 0x0010
};

enum
{
    MPG321_VERBOSE_PLAY   = 0x00000001,
    MPG321_QUIET_PLAY     = 0x00000002,
    MPG321_REMOTE_PLAY    = 0x00000004,

    MPG321_USE_OSS        = 0x00000010,
    MPG321_USE_SUN        = 0x00000020,
    MPG321_USE_ALSA       = 0x00000040,
    MPG321_USE_ESD        = 0x00000080,
    MPG321_USE_ARTS       = 0x00000100,
    MPG321_USE_NULL       = 0x00000200,
    MPG321_USE_STDOUT     = 0x00000400,
    MPG321_USE_WAV        = 0x00000800,
    MPG321_USE_AU         = 0x00001000,
    MPG321_USE_CDR        = 0x00002000,
    MPG321_USE_USERDEF    = 0x00004000,
    MPG321_USE_ALSA09     = 0x00008000,
    
    MPG321_FORCE_STEREO   = 0x00010000,
    MPG321_USE_SCROBBLER  = 0x00020000,
    MPG321_RECURSIVE_DIR  = 0x00040000,
    MPG321_PRINT_FFT	  = 0x00080000,
    MPG321_ENABLE_BASIC	  = 0x00100000,
    MPG321_ENABLE_BUFFER  = 0x01000000,
};

#define DEFAULT_PLAYLIST_SIZE 2048
#define BUF_SIZE 1048576 /* Size for read buffer for audio data */

/* playlist functions */
playlist * new_playlist();
void resize_playlist(playlist *pl);
char * get_next_file(playlist *pl, buffer *buf);
void add_cmdline_files(playlist *pl, char *argv[]);
void add_cmdline_files_recursive_dir(playlist *pl, char *argv[]);
void add_file(playlist *pl, char *file);
void load_playlist(playlist *pl, char *filename);
void set_random_play(playlist *pl);
void play_remote_file(playlist *pl, char *filename);
void clear_remote_file(playlist *pl);
void shuffle_files(playlist *pl);
void trim_whitespace(char *);

/* network functions */
int tcp_open(char * address, int port);
int udp_open(char * address, int port);
int raw_open(char * arg);
int http_open(char * arg);
int ftp_open(char * arg);

/* libmad interfacing functions */
enum mad_flow read_from_mmap(void *data, struct mad_stream *stream);
enum mad_flow read_from_fd(void *data, struct mad_stream *stream);
enum mad_flow read_header(void *data, struct mad_header const * header);
enum mad_flow output(void *data, struct mad_header const *header, struct mad_pcm *pcm);
int calc_length(char *file, buffer*buf );

static enum mad_flow handle_error(void *data, struct mad_stream *stream, struct mad_frame *frame);

enum mad_flow move(buffer *buf, signed long frames);
void seek(buffer *buf, signed long frame);
void pause_play(buffer *buf, playlist *pl);

/* libao interfacing and general audio-out functions */
void check_ao_default_play_device();
void check_default_play_device();
int playdevice_is_live();
void open_ao_playdevice(struct mad_header const *header);

/* remote control (-R) functions */
void remote_get_input_wait(buffer *buf);
enum mad_flow remote_get_input_nowait(buffer *buf);

/* options */
void parse_options(int argc, char *argv[], playlist *pl);

void mpg321_error(char *file);
void usage(char *);

RETSIGTYPE handle_signals(int sig) ;

/* xterm functions */
int tty_control();
int set_tty_raw();
int set_tty_restore();
int raw_print(char *ctlseq);
int osc_print(int ps1,int ps2,char *pt);
char *ctty_path();
void get_term_title(char *title);
int tty_read(char *output,size_t size);

/* AudioScrobbler functions */
void scrobbler_report(void);
void scrobbler_set_time(long);
void scrobbler_set_verbose(int);
RETSIGTYPE handle_sigchld(int sig);

/* FFT data structures */
#define FFT_BUFFER_SIZE_LOG 9
#define FFT_BUFFER_SIZE (1 << FFT_BUFFER_SIZE_LOG) /* 512 */
/*Temporary data stores to perform FFT in */
double real[FFT_BUFFER_SIZE];
double imag[FFT_BUFFER_SIZE];

typedef struct {
	double real[FFT_BUFFER_SIZE];
	double imag[FFT_BUFFER_SIZE];
} fft_state;

typedef short int sound_sample;
//void fft_perform(const sound_sample *input, double *output, fft_state *state);

fft_state *fft_init(void);

/*Basic control keys */
#define KEY_CTRL_VOLUME_UP	'*'
#define KEY_CTRL_VOLUME_DOWN	'/'
#define KEY_CTRL_NEXT_SONG	'n'
#define KEY_CTRL_PAUSE_SONG	'p'
#define KEY_CTRL_MUTE		'm'
/* This is it for the moment */


/* Output buffer process */
void frame_buffer_p();
/* Semaphore array */
int semarray;
/* Input/Output buffer position */
int mad_decoder_position;
int output_buffer_position;
/* Output Frame including needed information */
typedef struct {
	unsigned char data[4*1152];
	unsigned short length;
	signed long seconds;
	char time[80];
	unsigned long num_frames;
	struct mad_header header;
} output_frame;
/* Control structure with detailed information about each frame */
typedef struct {
	unsigned long total_decoded_frames;
	unsigned long total_played_frames;
	int done;
	int is_http;
	int is_file;
	int quit_now;
	int stop_playing_file;
	int timer;
	long bvolume;
} decoded_frames;

/* Output frame queue pointer */
output_frame *Output_Queue;

/* Shared total decoded frames */
decoded_frames *Decoded_Frames;

#if defined(__GNU_LIBRARY__) && !defined(_SEM_SEMUN_UNDEFINED)
/* */
#else
union semun {
	int val;
	struct semid_ds *buf;
	unsigned short *array;
	struct seminfo *__buf;
};
#endif
/* Get the next encoding/decoding place in the frame buffer */
int getnext_place(int position);

#endif /* _MPG321_H_ */


fbuffer.c
/*
    mpg321 - a fully free clone of mpg123.
    fbuffer.c: Copyright (C) 2011-2012 Nanakos Chrysostomos
    
    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#include <stdio.h>
#include <sys/types.h>
#include <unistd.h>
#include <signal.h>
#include <ao/ao.h>
#include <sys/sem.h>

#include "mpg321.h"

ao_device *play_device = NULL;
extern int quit_now;
extern int stop_playing_file;
extern int buffer_size;
signed long total_played_frames = 0;

void output_sighandler (int sig)	{
	ao_close (play_device);
	ao_shutdown ();
	exit (EXIT_SUCCESS);
}

void frame_buffer_p(){
	static unsigned int rate = 0;
	static int channels = 0;
	struct mad_header *header = NULL;
	output_frame *oframe = NULL;
	static struct sembuf read_sops = {0, 1, 0};
	static struct sembuf write_sops = {1, -1, 0};
	static struct sembuf end_sops = {2, 1, 0};
	decoded_frames *dframes = NULL;

	dframes = Decoded_Frames;
	dframes->total_played_frames = 0;

	pid_t ppid;

	ppid = getppid ();
	output_buffer_position = 0;
	signal (SIGUSR1, output_sighandler);
	ao_initialize();
	if (check_default_play_device_buffer() == -1)	{
		fprintf(stderr,"Unable to open play device");
		kill (ppid, SIGUSR1);
		handle_signals (SIGUSR1);
	}
	while (1)	{
		semop (semarray, &write_sops, 1);
		oframe = (Output_Queue+output_buffer_position);
		header = &(oframe->header);
		if (!play_device)	{
			channels = MAD_NCHANNELS(header);
			rate = header->samplerate;
			play_device = open_ao_playdevice_buffer(header);
			if (!play_device)	{
				fprintf(stderr,"Unable to open play device");
				kill (ppid, SIGUSR1);
				handle_signals (SIGUSR1);
			}
		} else if ((channels != MAD_NCHANNELS(header) || rate != header->samplerate) && playdevice_is_live()){
			ao_close(play_device);
			channels = MAD_NCHANNELS(header);
			rate = header->samplerate;
			play_device = open_ao_playdevice_buffer(header);
			if (!play_device)	{
				fprintf(stderr,"Unable to open play device");
				kill (ppid, SIGUSR1);
				handle_signals (SIGUSR1);
			}
		}
		if ((options.opt & MPG321_REMOTE_PLAY))
			fprintf(stdout,"%s",oframe->time);
		else if( options.opt & MPG321_VERBOSE_PLAY )
		{
			if(dframes->timer > 0)
				fprintf(stdout,"Volume: %lu%%  %s\r",dframes->bvolume,oframe->time);
			else
				fprintf(stdout,"%s",oframe->time);
		}

		if(oframe->num_frames <= 0 || dframes->quit_now || dframes->stop_playing_file || (dframes->is_http && dframes->done && (dframes->total_decoded_frames-total_played_frames)<=1) 
				|| (dframes->is_file && oframe->seconds <=0 && dframes->done) || (dframes->is_file && dframes->done && (dframes->total_decoded_frames-1 == total_played_frames)) 
				|| (dframes->is_http && dframes->done && (dframes->total_decoded_frames-1 == total_played_frames)) || (dframes->done && dframes->total_decoded_frames == 0) )
		{
			union semun sem_ops;
			sem_ops.val = buffer_size-1;
			semctl(semarray,0,SETVAL,sem_ops);
			sem_ops.val = 0;
			semctl(semarray,1,SETVAL,sem_ops);
			output_buffer_position = 0;
			if(stop_playing_file) stop_playing_file = 0;
			if(quit_now) quit_now = 0;
			dframes->total_decoded_frames = 0;
			dframes->done = 0;
			dframes->is_http = 0;
			dframes->is_file = 0;
			dframes->stop_playing_file = 0;
			total_played_frames = 0;
			semop(semarray,&end_sops,1);
			continue;
		}
		ao_play(play_device, oframe->data, oframe->length);
		total_played_frames+=1;
		dframes->total_played_frames += 1;
		output_buffer_position = getnext_place(output_buffer_position);
		semop (semarray, &read_sops, 1);
			
		if(dframes->timer > 0)
		{
			dframes->timer++;
			if(dframes->timer > 40)
			{
				
				if(!(options.opt & MPG321_VERBOSE_PLAY))
					fprintf(stdout,"                \r");
				dframes->timer = -1;
				fflush(stdout);
			}
		}

	}
}


options.c
/*
    mpg321 - a fully free clone of mpg123.
    options.c: Copyright (C) 2001, 2002 Joe Drew
    	       Copyright (C) 2006, 2007, 2008, 2009, 2010, 2011 Nanakos Chrysostomos
    Originally based heavily upon:
    plaympeg - Sample MPEG player using the SMPEG library
    Copyright (C) 1999 Loki Entertainment Software
    
    Also uses some code from
    mad - MPEG audio decoder
    Copyright (C) 2000-2001 Robert Leslie
    
    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif


#include "mpg321.h"
#include "getopt.h"

#include <sys/time.h>
#include <sys/resource.h>
#include <unistd.h>
#include <string.h>


/* Nanakos Chrysostomos (c) Copyright 2006 (nanakos@wired-net.gr)*/
extern int proxy_enable;
extern char *proxy_server;
extern int auth_enable;
extern int auth_enable_var;
extern remote_restart;
/*Basic Authentication Arguments*/
char *basic_auth = NULL;

/*xterm title setting flag*/
extern int set_xterm;

void parse_options(int argc, char *argv[], playlist *pl)
{
    struct option long_options[] =
    {
        /* NO-OPS. Implement these if you need them, please. */
    
        /* these take no parameter and have no short equiv */
        { "headphones", 0, 0, 'O' }, /* or -o h */
        { "speaker", 0, 0, 'P' }, /* or -o s */
        { "lineout", 0, 0, 'L' }, /* or -o l */
        { "reopen", 0, 0, 'N' },
        { "equalizer", 0, 0, 'E' },
        { "aggressive", 0, 0, 'I' },
        { "8bit", 0, 0, '8' },
    
        /* these take no parameter and have short equiv */
        { "2to1", 0, 0, '2' },
        { "4to1", 0, 0, '4' },
        { "check", 0, 0, 'c' },
        { "resync", 0, 0, 'y' },
        { "left", 0, 0, '0' },
        { "single0", 0, 0, '0' },
        { "right", 0, 0, '1' },
        { "single1", 0, 0, '1' },
        { "mono", 0, 0, 'm' },
        { "mix", 0, 0, 'm' },
        { "control", 0, 0, 'C' },
    
        /* these take a parameter and have short equiv */
        { "doublespeed", 1, 0, 'd' },
        { "halfspeed", 1, 0, 'h' },
        { "scale", 1, 0, 'f' },
        { "buffer", 1, 0, 'b' },
        { "proxy", 1, 0, 'p' },
        { "auth", 1, 0, 'u' },
        { "authvar", 1, 0, 'U' },
        { "rate", 1, 0, 'r' },
            
        /* The following are all implemented. */

        /* These take a parameter and have no short equiv */
        { "au", 1, 0, 'A' },
        { "cdr", 1, 0, 'D' },
    
        /* Takes no parameters */
        { "verbose", 0, 0, 'v' },
        { "quiet", 0, 0, 'q' }, 
        { "test", 0, 0, 't' },
        { "stdout", 0, 0, 's' },
        { "STDOUT", 0, 0, 's' },
        { "version", 0, 0, 'V' },
        { "help", 0, 0, 'H' },
        { "longhelp", 0, 0, 'H' },
        { "shuffle", 0, 0, 'z' },
        { "random", 0, 0, 'Z' },
        { "remote", 0, 0, 'R' },
        { "set_xterm", 0, 0, 'x' },
        { "stereo", 0, 0, 'T' },
        { "scrobbler", 0, 0, 'S' },
        { "recursive", 0, 0, 'B' },
        { "fft", 0, 0, 'F' },
        { "basickeys", 0, 0, 'K' },
        { "restart", 0, 0, '3' },
            
        /* takes parameters */
        { "frames", 1, 0, 'n' },
        { "skip-printing-frames", 1, 0, 'G' },
        { "output", 1, 0, 'o' },
        { "list", 1, 0, '@' },
        { "skip", 1, 0, 'k' },
        { "wav", 1, 0, 'w' },
        { "audiodevice", 1, 0, 'a' },
        { "gain", 1, 0, 'g' },
        { "loop", 1, 0, 'l' },
        { 0, 0, 0, 0 }
    };
    int option_index = 0, c;

    options.maxframes=-1;

    while ((c = getopt_long(argc, argv, 
                                "OPLTNEI824cy01mCd:h:f:b:r:G:" /* unimplemented */
                                "A:D:SBKFvqtsVHzZRxo:n:@:k:w:a:g:l:p:u:U:3",     /* implemented */
                        long_options, &option_index)) != -1)
    {            
        switch(c)
        {
            case 'O': case 'P': case 'L': case 'N': case 'E': case '8':
            case '2': case '4': case 'c': case 'y': case '0': case '1': case 'm': case 'C':
            case 'd': case 'h': case 'f': case 'b': 
                break;
	    case 'x':
		set_xterm=1;
		break;
	    case 'p':
		proxy_server = strdup(optarg);
		proxy_enable = 1;
		break;
	    case 'u':
		basic_auth = strdup(optarg);
		auth_enable = 1;
		auth_enable_var = 0;
		break;
	    case 'U':
		basic_auth = strdup(optarg);
		auth_enable_var = 1;
		auth_enable = 0;
            case 'n': 
                options.maxframes = atol(optarg);
                break;
            case 'r':
                break;

            case 'z':
                shuffle_play = 1;
                break;
                
            case 'Z':
                shuffle_play = 1;
                set_random_play(pl);
                break;
                
            case '@': 
                playlist_file = strdup(optarg);
                break;

            case 'v':
                options.opt |= MPG321_VERBOSE_PLAY;
		scrobbler_set_verbose(1);
                setvbuf(stdout, NULL, _IONBF, 0);
                break;
            
            case 'q':
                options.opt |= MPG321_QUIET_PLAY;
                break;
            
            case 'R':
                options.opt |= MPG321_REMOTE_PLAY;
                options.opt |= MPG321_QUIET_PLAY; /* surpress other output */
                setvbuf(stdout, NULL, _IONBF, 0);
                break;
                
            case 'k':
                options.seek = atol(optarg);
                status = MPG321_SEEKING;
                if (options.seek < 0)
                {
                    fprintf(stderr, "Number of frames to skip must be positive!\n");
                    exit(1);
                }
                break;
            
            case 't':
                options.opt = MPG321_USE_NULL;
                break;
            
            case 'w':
                options.opt |= MPG321_USE_WAV;
                options.device = strdup(optarg);
                break;
            
            case 'A':
                options.opt |= MPG321_USE_AU;
                options.device = strdup(optarg);
                break;
            
            case 'D':
                options.opt |= MPG321_USE_CDR;
                options.device = strdup(optarg);
                break;
            
            case 'a':
                /* use this device or file or whatever for output, with the
                   default output device, or whatever device is specified with -o */
                options.device = strdup(optarg);
                break;
            
            case 's':
                options.opt |= MPG321_USE_STDOUT;
                break;

	    case 'S':
		options.opt |= MPG321_USE_SCROBBLER;
		break;

	    case 'B':
		options.opt |= MPG321_RECURSIVE_DIR;
		break;

	    case 'F':
		options.opt |= MPG321_PRINT_FFT;
		break;
                
	    case 'K':
		options.opt |= MPG321_ENABLE_BASIC;
		break;

            case 'o':
                if (strcmp(optarg, "alsa") == 0)
                {
                    options.opt |= MPG321_USE_ALSA;
                }
                
                else if (strcmp(optarg, "alsa09") == 0)
                {
                    options.opt |= MPG321_USE_ALSA09;
                }
                
                else if (strcmp(optarg, "esd") == 0)
                {
                    options.opt |= MPG321_USE_ESD;
                }
                
                else if (strcmp(optarg, "arts") == 0)
                {
                    options.opt |= MPG321_USE_ARTS;
                }
                
                else if (strcmp(optarg, "oss") == 0)
                {
                    options.opt |= MPG321_USE_OSS;
                }

                else if (strcmp(optarg, "sun") == 0)
                {
                    options.opt |= MPG321_USE_SUN;
                }

                else if (strcmp(optarg, "h") == 0 || strcmp(optarg, "s") == 0 
                            || strcmp(optarg, "l") == 0)
                {
                    /* for now, we don't support these */
                }
                
                else /* Just pass on what the user gave to libao */
                {
                    options.opt |= MPG321_USE_USERDEF;
                    options.devicetype = strdup(optarg);
                }
                
                break;    
            
            case 'g':
                options.volume = mad_f_tofixed(atoi(optarg)/100.0);
		break;

            case 'l':
             loop_remaining = atoi(optarg);
              if (!loop_remaining)
                 loop_remaining = -1;
              break;

            case 'I':
                /* We only try to get high priority. If it fails (just as
                   mpg123 does when run as a user), that's fine */
                setpriority(PRIO_PROCESS, getpid(), -20);
                break;

            case 'T':
                options.opt |= MPG321_FORCE_STEREO;
                break;

            case 'G':
                options.skip_printing_frames = atoi(optarg);
                break;
	    case '3':
		remote_restart = 1;
		break;
            case 'V':
                printf("mpg321 version " VERSION ". Copyright (C) 2001, 2002 Joe Drew,\n"
		       "now maintained by Nanakos Chrysostomos and others.\n\n"
                       "This program is free software; you can redistribute it and/or modify\n"
                       "it under the terms of the GNU General Public License as published by\n"
                       "the Free Software Foundation; either version 2 of the License, or\n"
                       "(at your option) any later version.\n" );
                exit(0);
            
            case 'H':
                usage(argv[0]);
                exit(0);
            
            case ':':
                fprintf(stderr, "Missing argument to %s\n", argv[optind]);
                break;
        }
    }
}
